### PR TITLE
Enable PVC as DataSource for PVC creation

### PR DESCRIPTION
**Pull request by j-griffith**
This enables the ability to specify and existing PVC as a DataSource in
a new PVC Spec (eg "clone" and existing volume).

**What type of PR is this?**

> /kind feature

**What this PR does / why we need it**:

Currently the only valid input Kind for PVC DataSource is "VolumeSnapshot", this feature adds the ability to also specify a DataSource Kind of "PersistentVolumeClaim".

This results in the ability for a user to specify intent to clone and existing PVC, it does NOT implement cloning functionality in Kubernetes; that's left to the external csi provisioner and plugins.  It does however provide a mechanism to request a clone operation to the csi-provisioner and ultimately the csi plugin that the original PVC is allocated from.

Currently (as of 1.14) a user is able to enter a DataSource with any information they wish, however if it's not of the type "VolumeSnapshot" and the "VolumeSnapshotDataSource" feature gate is not enabled, the input is nilled and silently ignored.  This change follows the same pattern, via  the "VolumeDataSource" feature gate.

**Which issue(s) this PR fixes**:

Implements [enhancement/KEP 642](https://github.com/kubernetes/enhancements/pull/642)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
This change enables a user to specify a DataSource/Kind of type "PersistentVolumeClaim" in their PVC spec.  This can then be detected by the external csi-provisioner and plugins if capable.
```

### CODE REVIEW

File: pkg/api/persistentvolumeclaim/util.go [hunk 1]
```diff
@@ -52,3 +52,17 @@ func volumeSnapshotDataSourceInUse(oldPVCSpec *core.PersistentVolumeClaimSpec) b
 	}
 	return false
 }
+
+func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
+	if pvcSpec.DataSource != nil {
+		if pvcSpec.DataSource.Kind == "PersistentVolumeClaim" && utilfeature.DefaultFeatureGate.Enabled(features.VolumeDataSource) {
```
Comment (msau42) [inline pkg/api/persistentvolumeclaim/util.go]: These checks should also validate the Group

File: pkg/api/persistentvolumeclaim/util.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func DropDisabledFields(pvcSpec, oldPVCSpec *core.PersistentVolumeClaimSpec) {
 	if !utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) && !volumeModeInUse(oldPVCSpec) {
 		pvcSpec.VolumeMode = nil
 	}
-	if !utilfeature.DefaultFeatureGate.Enabled(features.VolumeSnapshotDataSource) && !volumeSnapshotDataSourceInUse(oldPVCSpec) {
+	if !dataSourceIsEnabled(pvcSpec) && !dataSourceInUse(oldPVCSpec) {
```
Comment (msau42) [inline pkg/api/persistentvolumeclaim/util.go]: can you add unit tests for this

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
 	// This is a beta feature.
 	// +optional
 	VolumeMode *PersistentVolumeMode
-	// This field requires the VolumeSnapshotDataSource alpha feature gate to be
-	// enabled and currently VolumeSnapshot is the only supported data source.
-	// If the provisioner can support VolumeSnapshot data source, it will create
-	// a new volume and data will be restored to the volume at the same time.
-	// If the provisioner does not support VolumeSnapshot data source, volume will
+	// This field can be used to specify either:
+	// * An existing VolumeSnapshot
```
Comment (msau42) [inline pkg/apis/core/types.go]: Be specific about the exact object types supported, ie "snapshot.storage.k8s.io/VolumeSnapshot"

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
...
+	// * An existing Volume (PVC, Clone operation)
+	// In order to use either of these DataSource types, the appropriate feature gate
+	// must be enabled (VolumeSnapshotDataSource, VolumeDataSource)
+	// If the provisioner can support the specified data source, it will create
+	// a new volume based on the contents of the specified PVC or Snapshot.
+	// If the provisioner does not support the specified data source, the volume will
 	// not be created and the failure will be reported as an event.
 	// In the future, we plan to support more data source types and the behavior
```
Comment (msau42) [inline pkg/apis/core/types.go]: Add a note that both of these sources are alpha

File: pkg/apis/core/validation/validation.go [hunk 5]
```diff
@@ -1524,6 +1524,7 @@ var supportedVolumeModes = sets.NewString(string(core.PersistentVolumeBlock), st
 
 var supportedDataSourceAPIGroupKinds = map[schema.GroupKind]bool{
 	{Group: "snapshot.storage.k8s.io", Kind: "VolumeSnapshot"}: true,
+	{Group: "", Kind: "PersistentVolumeClaim"}:                 true,
```
Comment (msau42) [inline pkg/apis/core/validation/validation.go]: I think core objects should still have a group

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -416,11 +416,14 @@ type PersistentVolumeClaimSpec struct {
...
-	// a new volume and data will be restored to the volume at the same time.
-	// If the provisioner does not support VolumeSnapshot data source, volume will
+	// This field can be used to specify either:
+	// * An existing VolumeSnapshot
+	// * An existing Volume (PVC, Clone operation)
+	// In order to use either of these DataSource types, the appropriate feature gate
+	// must be enabled (VolumeSnapshotDataSource, VolumeDataSource)
+	// If the provisioner can support the specified data source, it will create
+	// a new volume based on the contents of the specified PVC or Snapshot.
+	// If the provisioner does not support the specified data source, the volume will
```
Comment (msau42) [inline pkg/apis/core/types.go]: What happens if the provisioner is on an older version that doesn't understand the new field?

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (j-griffith) [reply pkg/apis/core/validation/validation.go]: @msau42 It seemed redundant to require it based on the definition here:  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L4062-L4065

The object checks default to the core type so it's not needed for any of the implementation.

File: pkg/apis/core/types.go (hunk 6 above)
Comment (j-griffith) [reply pkg/apis/core/types.go]: If that's the case then the current CSI Provisioner includes a validation check of it's own and it fails provision for any DataSource Kind except Snapshot.  So the result would just be an error reported back from provisioning, the requested PVC should then go into error status.

File: pkg/apis/core/types.go (hunk 3 above)
Comment (j-griffith) [reply pkg/apis/core/types.go]: Sure, I'll reword this to explicitly call out the API

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (xing-yang) [reply pkg/apis/core/validation/validation.go]: Core API's group is "".  That is why we had to make APIGroup optional.

File: pkg/features/kube_features.go [hunk 7]
```diff
@@ -434,6 +434,12 @@ const (
 	//
 	// Enables the OpenStack Cinder in-tree driver to OpenStack Cinder CSI Driver migration feature.
 	CSIMigrationOpenStack utilfeature.Feature = "CSIMigrationOpenStack"
+
+	// owner: @j-griffith
+	// alpha: v1.14
+	//
+	// Enable data source support (applies to all supported types, snapshots, populators and PVCs).
+	VolumeDataSource utilfeature.Feature = "VolumeDataSource"
```
Comment (xing-yang) [inline pkg/features/kube_features.go]: This probably should be "VolumePVCDataSource".

Comment (j-griffith) [reply pkg/features/kube_features.go]: I'm open to changing to that; I was just following the pattern s/VolumeSnapshot/Volume/ but if it seems more clear to folks I can change the name.

As far as the comment, I need to update that; that's from the original version where I wanted to combine the feature gates into a single option (decided not to fight that battle) so I'll fix the comment.

File: pkg/api/persistentvolumeclaim/util.go (hunk 2 above)
Comment (j-griffith) [reply pkg/api/persistentvolumeclaim/util.go]: Done

File: pkg/api/persistentvolumeclaim/util.go (hunk 1 above)
Comment (j-griffith) [reply pkg/api/persistentvolumeclaim/util.go]: Added to both PVC and Snapshot

File: pkg/features/kube_features.go [hunk 8]
```diff
@@ -436,10 +436,10 @@ const (
 	CSIMigrationOpenStack utilfeature.Feature = "CSIMigrationOpenStack"
 
 	// owner: @j-griffith
-	// alpha: v1.14
+	// alpha: v1.15
 	//
-	// Enable data source support (applies to all supported types, snapshots, populators and PVCs).
-	VolumeDataSource utilfeature.Feature = "VolumeDataSource"
+	// Enable support for specifying an existing PVC as a DataSource
+	VolumePVCDataSource utilfeature.Feature = "VolumePVCDataSource"
```
Comment (j-griffith) [inline pkg/features/kube_features.go]: I think the old name was more clear; but this works for me if other prefer it.

**Comment by j-griffith** [comment]
Sorry about the messiness; didn't squash the rebase commits and had around 420 commits added, tried to fix up by resetting to an older commit id and squashing from there.

**Comment by j-griffith** [comment]
Closing this out and pointing to updated version without the rebase commits that added every label in the project:  https://github.com/kubernetes/kubernetes/pull/76913
